<?xml version="1.0" ?>
<launch>
	<!-- If mode = sim: Simulated AUV with MBES and control in loaded bathymetric map
	If mode = gt: reproduce Hugin survey and construct map of real bathymetry -->
    <arg name="mode" default="sim"/>
	<arg name="namespace" default="hugin"/>
	<arg name="debug" default="0"/>
    <arg name="odom_rate" default="0.1"/>
	<arg name="meas_rate" default="0.5"/>
	<arg name="mbes_sim_threads" default="8"/>
    <arg name="map_name" default="lost_targets"/> <!-- Name of UFO map to be loaded (if exists) or saved when gt mission is over -->

    <!-- <arg name="path" default="/media/torroba18/Seagate Expansion Drive/HuginDatasets/Antarctica/NBP_19_02_011/pp/EM2040/dataset"/> -->
    <!-- <arg name="path" default="/media/torroba18/Seagate Expansion Drive/HuginDatasets/BornoSMaRC201906/20190626_Mission_8/pp/EM2040/KTH1"/> -->
	<arg name="path" default="/media/torroba18/Seagate Expansion Drive/MMTDatasets/LostTargets/Ping_Processed/5-Products/MBES_GSF/Post_Deployment_PROCESSED"/>
    <!-- <arg name="path" default="/media/torroba18/Seagate Expansion Drive/HuginDatasets/BornoSMaRC201906/2019_06_24_Biosensor/pp/EM2040/map_section"/> -->
	<!-- <arg name="path" default="/media/torroba18/Seagate Expansion Drive/HuginDatasets/BornoSMaRC201906/Mission_58_20190619_1/pp/EM2040"/> -->
	<!-- <arg name="path" default="/home/torroba18/Downloads/Faulty_mbes"/> -->
	<arg name="cereal_trajectory" default="$(arg path)/mbes_pings.cereal"/>



	<!-- TODO: set tf map to odom in a more intuitive way -->
	<!-- Coordinates of odom frame wrt the map frame. By default, the AUV start the mission here -->
	<arg name="x" default="-200.0"/>
	<arg name="y" default="-20.0"/>
	<arg name="z" default="0.0"/>
	<arg name="roll" default="0.0"/>
	<arg name="pitch" default="0.0"/>
	<arg name="yaw" default="3.0"/>

    <!--UFO mapping parameters-->
    <arg name="nodelet_manager" default="bathy_mapper_server_nodelet_manager" />
    <arg name="resolution" default="1" />
    <arg name="depth_levels" default="21" />
    <arg name="multithreaded" default="true" />
    <arg name="map_latch" default="true" />
	<arg name="map_binary_latch" default="false" />
	<arg name="map_cloud_latch" default="false" />

	<group ns="$(arg namespace)">

		  <!-- Particle filter -->
		<include file="$(find auv_particle_filter)/launch/auv_pf.launch" >
			<arg name="mode" value="$(arg mode)"/>
		</include>

		<!-- Hugin model -->
		<param name="robot_description" command="$(find xacro)/xacro.py '$(find hugin_description)/robots/hugin_auv_default.urdf.xacro' debug:=$(arg debug) namespace:=$(arg namespace)" />
		<node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="true" output="screen">
			<!-- <param name="robot_description" value="/$(arg namespace)/robot_description" /> -->
		</node>

		<!-- UFO mapping -->
	    <node pkg="nodelet" type="nodelet" name="$(arg nodelet_manager)" args="manager" output="screen">
            <param name="num_worker_threads" value="12" />
	    </node>

		<node pkg="nodelet" type="nodelet" name="bathy_mapper_server_nodelet" args="load bathy_mapper/BathyMapperNodelet $(arg nodelet_manager)" output="log" required="true">
	        <remap from="cloud_in" to="/gt/mbes_pings" />
	        <param name="output_map" value="$(find bathy_mapper)/maps/$(arg map_name)_map" />
	        <param name="input_map" value="$(find bathy_mapper)/maps/$(arg map_name)_map" />
	        <param name="multithreaded" value="$(arg multithreaded)" />
	        <param name="resolution" value="$(arg resolution)" />
	        <param name="depth_levels" value="$(arg depth_levels)" />
			<param name="odom_rate" value="$(arg odom_rate)" />
			<param name="meas_rate" value="$(arg meas_rate)" />
	        <param name="world_frame" value="world" />
			<param name="map_frame" value="map" />
			<param name="mbes_link" value="$(arg namespace)/mbes_link" />
			<param name="mbes_open_angle" value="1.5708" />
			<param name="num_beams_sim" value="30" />
			<param name="mbes_sim_as" value="/mbes_sim_server" />
		</node>

		<!-- Simulate Hugin AUV -->
		<group if="$(eval mode == 'sim')">
    		<node pkg="tf" type="static_transform_publisher" name="tf_world_map_pub" args="0 0 0 0 0 0 world map 50" />

    		<node pkg="tf" type="static_transform_publisher" name="tf_map_odom_pub" args="$(arg x) $(arg y) $(arg z) $(arg yaw) $(arg pitch) $(arg roll) map odom 50" />

			<node pkg="auv_model" type="auv_motion_model" name="auv_motion_model" output="screen">
				<param name="odom_rate" value="$(arg odom_rate)" />
				<param name="meas_rate" value="$(arg meas_rate)" />
				<param name="odom_sim" value="/$(arg mode)/odom" />
				<param name="world_frame" value="world" />
				<param name="map_frame" value="map" />
				<param name="odom_frame" value="odom" />
				<param name="base_link" value="$(arg namespace)/base_link" />
				<param name="mbes_link" value="$(arg namespace)/mbes_link" />
				<param name="throttle_cmd" value="/$(arg namespace)/throttle_cmd" />
				<param name="thruster_cmd" value="/$(arg namespace)/thruster_cmd" />
				<param name="inclination_cmd" value="/$(arg namespace)/inclination_cmd" />
				<param name="mbes_sim_as" value="/mbes_sim_server" />
			</node>

			<node pkg="auv_model" type="auv_sim_teleop.py" name="auv_sim_teleop" output="screen">
				<param name="throttle_cmd" value="/$(arg namespace)/throttle_cmd" />
				<param name="thruster_cmd" value="/$(arg namespace)/thruster_cmd" />
				<param name="inclination_cmd" value="/$(arg namespace)/inclination_cmd" />
			</node>
  		</group>

		<!-- Reproduce real Hugin survey -->
  		<group unless="$(eval mode == 'sim')">
			<node pkg="auv_2_ros" type="auv_2_ros" name="auv_2_ros" output="screen">
				<param name="trajectory_cereal" value="$(arg cereal_trajectory)"/>
				<param name="odom_rate" value="$(arg odom_rate)" />
				<param name="meas_rate" value="$(arg meas_rate)" />
				<param name="mbes_pings" value="/gt/mbes_pings" />
				<param name="map_pcl" value="/gt/map" />
				<param name="debug_pings" value="debug/mbes_pings" />
				<param name="sim_pings" value="/sim/mbes_pings" />
				<param name="world_frame" value="world" />
				<param name="map_frame" value="map" />
				<param name="odom_frame" value="odom" />
				<param name="base_link" value="$(arg namespace)/base_link" />
				<param name="mbes_link" value="$(arg namespace)/mbes_link" />
			</node>
		</group>

	</group>

</launch>
